gossip/src/cluster_info.rs

```diff
@@ -1243,6 +1283,29 @@ impl ClusterInfo {
             .collect()
     }
 
+    /// Returns last-voted-fork-slots inserted since the given cursor.
+    /// Excludes entries from nodes with unkown or different shred version.
+    pub fn get_restart_last_voted_fork_slots(
+        &self,
+        cursor: &mut Cursor,
+    ) -> Vec<(EpochSlotsIndex, EpochSlots, Slot, Hash)> {
+        let self_shred_version = Some(self.my_shred_version());
+        let gossip_crds = self.gossip.crds.read().unwrap();
+        gossip_crds
+            .get_restart_last_voted_fork_slots(cursor)
+            .filter(|entry| {
+                let origin = entry.value.pubkey();
+                gossip_crds.get_shred_version(&origin) == self_shred_version
```

Comment:
If we need this check, shouldn't shred version be included in the `RestartLastVotedForkSlots`?
There is no guarantee that the `RestartLastVotedForkSlots` and the `ContactInfo` you obtain shred-version from are in sync.

Reply:
@wen-coding do we need to include `shred_version` in `RestartLastVotedForkSlots`?

Reply:
This code was copied from get_epoch_slots, so I think all other Gossip messages which check whether shred_version matches would have the same problem, since we could always have a lag on ContactInfo shred-version.

So shred_version is just a sanity check, because it is so easy to circumvent in a real attack. But it's nice to have for naive users pointing mainnet validators at testnet or vice versa. I could go either way here, but would prefer to not include shred_version in RestartLastVotedForkSlots for brevity. Using a contact-info shred_version which may occasionally go out of sync is fine, we will drop some messages but can pick them up later I believe? For RestartLastVotedForkSlots since we draw a line between Gossip messages in restart and those not in restart, this check is less important.

Reply:
> This code was copied from get_epoch_slots

* get_epoch_slots may have been done improperly. copying the same mistakes will not help.
* For `EpochSlots` we are not expecting a shred-version change. Here there is always a shred-version change when the `RestartLastVotedForkSlots` is pushed or needed.

> I think all other Gossip messages which check whether shred_version matches would have the same problem

most of those which check shred-version are node's sockets which are obtained from the same contact-info message which has shred-version.

> For RestartLastVotedForkSlots since we draw a line between Gossip messages in restart and those not in restart, this check is less important.

I don't understand this, why?
My understanding is that for `RestartLastVotedForkSlots` it is even more important, because, again, there is always a shred-version change when the `RestartLastVotedForkSlots` is pushed or needed.

Reply:
Okay, maybe it is overstatement to say this check is less important.

The shred_version change was put in mainly to prevent RestartLastVotedForkSlots messages from occupying CRDS table space in non-restarting validators because they are big, but later we decided to add extra check "normal validators just don't accept RestartLastVotedForkSlots at all no matter what shred_version it has", so the shred_version change becomes an extra protection.

Let's say one validator has its contact info propagating slower than RestartLastVotedForkSlots, so some neighbors dropped its RestartLastVotedForkSlots because of shred_version mismatch. I was hoping newly restarted validators will do a Gossip pull for all messages and somehow help spread the lost RestartLastVotedForkSlots messages to others?

If that's not the case then it might be better to add shred_version in the messages themselves.